Carsten's comments from 2nd WGLC (#44)
core-wg/groupcomm-bis: draft-ietf-core-groupcomm-bis.md

In particular, the responsible OSCORE Group Manager registers its security groups to the RD, as links to its own corresponding resources for joining the security groups {{I-D.ietf-ace-key-groupcomm-oscore}}. Later on, CoAP endpoints can discover the names of the registered security groups and related application groups, by using the lookup interface of the RD, and then join the security group through the respective Group Manager.

> #### Discovery from the CoAP Servers of an Application/CoAP Group ### {#sssec-discovery-from-servers}

Reviewer: (Really a nit: Discovery "from X" could be discovery that originates from X or discovery that works by inquiring X.  No good suggestion here.)

Maybe clarify the unclear "of" of "Servers of a ... group" by including "members"?

```suggestion
#### Discovery from Server Members of an Application or CoAP Group ### {#sssec-discovery-from-servers}
```

Another method to more easily meet the above constraint is to instantiate multiple CoAP clients at multiple UDP ports on the same host. The Token values only have to be unique within the context of a single CoAP client, so using multiple clients can make it easier to meet the constraint.

### Client Handling of Multiple Responses With Same Token ### {#sec-request-response-multi}
Since a client sending a group request with a Token T will accept multiple responses with the same Token T, it is possible in particular that the same server sends multiple responses with the same Token T back to the client.

> For example, if the client sends a group request specifying the Observe option set to 0 (see {{Section 3.1 of RFC7641}}) and this server adds the client to the list of observers for the targeted resource, then the server will send multiple responses as Observe notifications to notify the client of changes to the resource state (see {{Section 4.2 of RFC7641}}). The use of Observe with group communication is discussed in more details in {{sec-observe}}. As an alternative example, this server might not implement the optional CoAP message deduplication based on Message ID; or it might be acting out of specification as a malicious, compromised or faulty server.

Reviewer: ```suggestion
For example, if the client sends a group request specifying the Observe option set to 0 (see {{Section 3.1 of RFC7641}}) and this server adds the client to the list of observers for the targeted resource, then the server is set up to send multiple responses as Observe notifications to notify the client of changes to the resource state (see {{Section 4.2 of RFC7641}}). The use of Observe with group communication is discussed in more details in {{sec-observe}}. As another example, a server might not implement the optional CoAP message deduplication based on Message ID; or it might be acting out of specification as a malicious, compromised or faulty server.
```

When this happens, it is up to the specific client implementation to decide at which layer deduplication of responses is performed, or whether it is necessary in an application at all. If the processing of a response is successful, the client delivers the response to the application as usual.

> The application itself can be in a good position to decide what to do, depending on the available context information. For instance, it might accept and process all the responses from the same server, even if they are not Observe notifications (i.e., they do not include an Observe option). Alternatively, the application might accept and process only one of those responses, such as the one that is deemed to be the most recent one from that server, e.g., when this can trigger a change of state within the application.

Reviewer: My problem still is that, from a client's point of view, each incoming response message (unless detected to be reordered) will be deemed to be the most recent one from that server, so I don't understand how this statement is selective.

Author: Maybe it's just better to drop the central part of the sentence and say:

> Alternatively, the application might accept and process only one of those responses, e.g., when this can trigger a change of state within the application.

Author: Addressed in https://github.com/core-wg/groupcomm-bis/pull/44/commits/f58845b1f2f542506ddf1c6b8479748fa9599e07

A server endpoint that includes an ETag Option in a response to a group request SHOULD construct the ETag Option value in such a way that the value
will be unique to this particular server with a high probability. This practically prevents a collision of the ETag values from different servers in the same CoAP group and application group, which in turn allows the client to effectively validate a particular response of an origin server. This can be accomplished, for example, by embedding a compact ID (or hash) of the server within the ETag value, where the ID is unique (or unique with a high probability) in the scope of the CoAP/application groups.

> Note: a CoAP server implementation that is unaware of the updates to {{RFC7252}} made by this document will expect group requests to never contain an ETag Option (see {{Section 8.2.1 of RFC7252}}). Such a server treats an ETag Option in a group request as an unrecognized option per {{Sections 5.4 and 8.2.1 of RFC7252}}, causing it to ignore this (elective) ETag Option regardless of its value, and process the request normally as if that ETag Option was not included.

Reviewer: ```suggestion
Note: a CoAP server implementation that is unaware of the updates to {{RFC7252}} made by this document will expect group requests to never contain an ETag Option (see {{Section 8.2.1 of RFC7252}}). Such a server treats an ETag Option in a group request as an unrecognized option per {{Sections 5.4 and 8.2.1 of RFC7252}}, causing it to ignore this (elective) ETag Option regardless of its value, and processes the request normally as if that ETag Option was not included.
```

## URI Path Selection ##
The URI Path used in a group request is preferably a path that is known to be supported across all members of a CoAP group. However, there are valid use cases where a group request is known to be successful only for a subset of the CoAP group. For instance, the subset may include only members of a specific application group, while the members of the CoAP group for which the request is unsuccessful (for example because they are outside the target application group) either suppress a response as per the default behavior from {{sec-request-response-suppress}}, or reply with an error response, e.g., when the default behavior is overridden by a No-Response Option {{RFC7967}} included in the group request.

## Port Selection for UDP Transport ##

> A server that is a member of a CoAP group listens for CoAP request messages on the group's IP multicast address and port number. The group's port number is usually the CoAP default UDP port number 5683, or alternatively another non-default UDP port number if configured. Regardless of the method that is used for selecting the group's port number, the same port number is used across all CoAP servers that are members of a CoAP group and across all CoAP clients sending group requests to that group.

Reviewer: Maybe be terminally clear which of the port numbers we are talking about:

```suggestion
A server that is a member of a CoAP group listens for CoAP request messages on the group's IP multicast address and port number. The group's port number is usually the CoAP default UDP port number 5683, or alternatively another non-default UDP port number if configured. Regardless of the method that is used for selecting the group's port number, the same port number is used as the destination port number for requests across all CoAP servers that are members of a CoAP group and across all CoAP clients sending group requests to that group.
```

A client that has sent a group GET or FETCH request with the Observe Option MAY follow up by sending a new unicast CON request with the same Token value and same Observe Option value to a particular server, in order to ensure that the particular server receives the request. This is useful in case a specific member of the CoAP group did not respond to the initial group request, although it was expected to. In this case, the client MUST use a Message ID that differs from that of the initial group request message.

> Since the first Observe notification from a server can be lost, a client SHOULD be ready to begin receiving the Observe notifications from a server long after the Non-confirmable group request with the Observe Option was sent.

Reviewer: (Another unqualified SHOULD, but maybe this is clear from the context here.)

Author: Yes, that was my hope.

Author: Marking this point as resolved, as considered clear from the context.

{{Section 2.8 of RFC7959}} specifies how a client can use block-wise transfer (Block2 Option) in a multicast GET request to limit the size of the initial response of each server. Consistent with {{Section 2.5 of RFC8132}}, the same can be done with a multicast FETCH request.

If a client sends a multicast GET or FETCH request including a Block2 Option with a block number of 0, then the client can rely on two possible approaches in order to retrieve any further blocks of the resource from responding servers.

1. The client uses unicast requests, separately addressing each different server.

> 2. The client uses follow-up group requests, if all the responses received from different servers specify the same block size in their Block2 Option. In particular, such a block size can be equal to the block size specified in the Block2 Option of the first group request, or instead a smaller one. If the client relies on this approach, then the Block2 Option of follow-up group requests in the same block-wise transfer MUST specify the same block size used by all the servers in the Block2 Option of their responses.

Reviewer: I'm still not sure the MUST is justified here (or is it intended as a restatement of something I'm missing?)

Author: It's a restatement of what is defined in Section 2.4 of of RFC 7959 (of course, there discussed for one-to-one communication). The same goes for the other "MUST" two paragraphs below.

https://github.com/core-wg/groupcomm-bis/pull/44/commits/d26b390a23cd2fe8816d872e32ef7ac3d4246e2a  addressed this by avoiding "MUST" and just asserting, i.e., s/MUST specify/specifies

Author: Right, that should be the latter :-)

Author: Marking this point as resolved, per the discussion and commit above

### TCP, TLS, and WebSockets ###

CoAP over TCP, TLS, and WebSockets is defined in {{RFC8323}}. Although it supports unicast only, it can be employed as a transport for CoAP group communication in situations where unicast is used, such as exchanging messages with a proxy and completing block-wise transfers. In particular:

* A suitable cross-proxy can be set up, such that it receives a unicast CoAP group request over TCP/TLS/WebSockets, and then forwards the request to the servers in the group over UDP/IP multicast (see {{sec-proxy}}). The discovery of such a proxy can rely on means defined in {{I-D.ietf-core-transport-indication}}.

* {{RFC8323}} can be employed to complete block-wise transfers for CoAP group communication, with the limitations discussed in {{sec-block-wise}}.

>   That is, after the first group request including the Block2 Option and sent over UDP, the following unicast CoAP requests targeting individual servers to retrieve further blocks may be sent over TCP or WebSockets, possibly protected with TLS.

Reviewer 2: On "possibly protected with TLS" -> if this is done, then the client would not be accessing the same resource anymore right? 

Because the resource namespaces of schemes coap:// and coaps:// are separately defined and not necessarily identical, as I recall it.

Author: The hard namespace separation is indeed what was originally intended in Section 6.2 of RFC 7252 and Section 8 of RFC 8323.

(Note that following-up with unprotected TCP traffic would still mean a change of scheme compared to the first request, i.e., from coap to coap+tcp)

Fundamentally, yes, the block-wise transfer has to continue for the same resource.

When considering servers reached individually, I think that the key here was about:

> This requires the individually addressed servers ... to also support CoAP over TCP/TLS/WebSockets **for the targeted resource**. ...

> ... discovering that a server has those transports available and that they allow **accessing the targeted resource**, ... . Such discovering can rely on means defined in {{I-D.ietf-core-transport-indication}}.

@chrysn  , what do you think?

Reviewer: This is a problem that should be solved by transport indication.

Reviewer 2: @marco-tiloca-sics Right, we do say "the targeted resource" explicitly. So it does look like we can transfer the problem entirely to the {{I-D.ietf-core-transport-indication}} which is already referenced.

(Assuming the discovery problem is solved in there - haven't studied it in detail.)

So we can leave the current text as is, I think.

### TCP, TLS, and WebSockets ###

> Because it supports unicast only, employing {{RFC8323}} (CoAP over TCP, TLS, and WebSockets) as a transport for CoAP group communication is limited to the use of block-wise transfer discussed in {{sec-block-wise}}.

Reviewer: Thank you for improving the wording.
Couldn't CO transports still be used as the first hop to a proxy that then uses IP multicast?

Author: Yes. What about adding the following new paragraph at the end of the same Section 3.9.4 "TCP, TLS, and WebSockets" ?

> Alternatively, a suitable cross-proxy can be set up, such that it receives a unicast CoAP group request over TCP/TLS/WebSockets, and then forwards the request to the servers in the group over IP multicast (see {{sec-proxy}}). The discovery of such a proxy can also rely on means defined in {{I-D.ietf-core-transport-indication}}.

Reviewer: I'm not sure there is an "alternative" here.  I would just say that 8323 is useful in unicast situations, such as

* completing block-wise
* talking to a proxy

Author: Thanks!

Now in https://github.com/core-wg/groupcomm-bis/pull/44/commits/2327d509f69153ed9c23fcc8aaf3542bae2fadfa  , where the main point is stated upfront. @cabo @EskoDijk , please have a look.

Reviewer: The RFC editor will provide some minor wording improvements, so I think we can ship this now.

Reviewer 2: @marco-tiloca-sics I'm also okay with the new text - better now.

This requires a server to have a Replay Window that is in a valid state. If the server's Replay Window is initialized as invalid, e.g., due to a reboot, the server should use the challenge-response synchronization method based on the Echo Option for CoAP defined in {{RFC9175}} as described in {{Section 9 of I-D.ietf-core-oscore-groupcomm}}, in order to make the Replay Window valid before resuming to accept incoming messages from other group members.

## Use of CoAP No-Response Option ##

When CoAP group communication is used in CoAP NoSec (No Security)
mode (see {{chap-unsecured-groupcomm}}), the CoAP No-Response Option {{RFC7967}} could be misused by a malicious client to evoke as many responses from servers to a group request as possible, by using the value '0' -- Interested in all responses. This might even override the default behavior of a CoAP server to suppress the response in case there is nothing of interest to respond with. Therefore, this option can be used to perform an amplification attack (see {{ssec-amplification}}).

> As a mitigation, a server that takes into account the No-Response Option relaxes the standard suppression rules for a resource only in case the option is sent by an authenticated client. Consequently, if sent by an unauthenticated client, the option can still be used to expand the classes of responses suppressed compared to the default rules, but not to reduce the classes of responses suppressed.

Reviewer: Not sure that indicative mood ("relaxes") works here.

» In grammar, the indicative mood is a verb mood that the speaker or writer uses to express information that sounds factual. «

Author: Changed to "can specifically relax"

> As a mitigation, a server that takes into account the No-Response Option SHOULD relax the standard suppression rules for a resource only in case the option is sent by an authenticated client. Consequently, if sent by an unauthenticated client, the option can still be used to expand the classes of responses suppressed compared to the default rules, but not to reduce the classes of responses suppressed.

Reviewer: This has a SHOULD without explanation what the exceptions are.
Isn't this really just a statement about when the mitigation actually works?

With the exception of such particular use cases, group communications MUST be secured using Group OSCORE {{I-D.ietf-core-oscore-groupcomm}}, see {{chap-oscore}}. As discussed in {{chap-security-considerations-sec-mode-attacks}}, this limits the feasibility and impact of amplification attacks.

## Replay of Group Requests ##

After a client has sent a group request over IP multicast, an adversary might capture the group request to be re-injected in the group as a replay to the server nodes. In particular:

> * If the adversary re-injects the group request before the client has freed up the corresponding Token value (see {{sec-token-reuse}}), the client might receive additional responses from the servers in the group.

Reviewer 2: This is only from the servers that do *not*  implement the optional deduplication based on Message ID. Should we clarify that?   In cases where all servers implement deduplication, this replay will have no impact.

Reviewer: Message-ID deduplication is time-limited, so an attacker can circumvent that by just waiting long enough.  But yes, that is another thing an attacker has to consider.

Reviewer 2: Looking at the context of this section within the document: it's about replay for both protected and unprotected group requests.

If unprotected, the attacker could just replay with a changed MID - so the attacker doesn't even have to wait for 145 seconds.   The attacker could just forge a completely new request also, so why even replay :)
Not sure if there's a lot of value going into the replay details for this case.

For the Group OSCORE case it says that it's protected against these types of replay.

Reviewer: The current text overstates the mitigating effect of MID dedup.  I think we need to roll this back a bit and then we can ship.

Reviewer 2: Ok for me to partially roll it back. Text can become for example:

```
* If the adversary re-injects the group request before the client has freed up the corresponding Token value (see {{sec-token-reuse}}), the client might receive additional responses from one or more of the servers in the group.
```

The "one or more" can cover various cases where some servers implement deduplication, others not. And cases where despite deduplication for example the client's Token-valid time is longer than 145 sec, and the attacker replays the message after 145 sec.   Also cases where the attacker modifies the MID to circumvent the deduplication could be considered included, as a form of "almost replay". 

For me it's also ok to fully roll back the change. The phrase "might receive additional responses from the servers in the group" doesn't say how many servers would respond: it could be 0, 1, N or all.

Author: Thanks! The suggested short version is now in https://github.com/core-wg/groupcomm-bis/pull/44/commits/11ec9f2277f7b65bb0ca085fbbc95e7c747a3f07

After a client has sent a group request over IP multicast, an adversary might capture the group request to be re-injected in the group as a replay to the server nodes. In particular:

> * If the adversary re-injects the group request before the client has freed up the corresponding Token value (see {{sec-token-reuse}}), the client might receive additional responses from the servers in the group, i.e., from the the servers that do not implement the optional Message ID based deduplication ({{Section 4.5 of RFC7252}}).

Reviewer: As I said, MID dedup times out, so the condition is too simplistic here.

s/the the/the/

Author: As in https://github.com/core-wg/groupcomm-bis/pull/44/commits/11ec9f2277f7b65bb0ca085fbbc95e7c747a3f07 , the MID deduplication is now not mentioned anymore.

@cabo , can we simply keep it that way?

Alternatively, a possible additional sentence that provides some explanation can be (highlighted in bold):

> If the adversary re-injects the group request before the client has freed up the corresponding Token value (see {{sec-token-reuse}}), the client might receive additional responses from one or more of the servers in the group. **For instance, a server might respond if it does not implement the optional Message ID based deduplication, or more generally if at least NON_LIFETIME seconds elapse between receiving the original group request and the re-injected one (see {{Section 4.5 of RFC7252}}).**

Author: After re-checking this, we can better keep it **without** the new sentence, to avoid restating.
